In [1]:
import numpy as np
import pandas as pd
pd.set_option("display.max_columns", None)
pd.set_option('display.max_rows', None)

In [2]:
bank_df = pd.read_csv("../data/raw/bank-additional.csv")
bank_df.head()

Unnamed: 0   age        job  marital    education  default  housing  loan  \
0           0   NaN  housemaid  MARRIED     basic.4y      0.0      0.0   0.0   
1           1  57.0   services  MARRIED  high.school      NaN      0.0   0.0   
2           2  37.0   services  MARRIED  high.school      0.0      1.0   0.0   
3           3  40.0     admin.  MARRIED     basic.6y      0.0      0.0   0.0   
4           4  56.0   services  MARRIED  high.school      0.0      0.0   1.0   

     contact  duration  campaign  pdays  previous     poutcome  emp.var.rate  \
0  telephone       261         1    999         0  NONEXISTENT           1.1   
1  telephone       149         1    999         0  NONEXISTENT           1.1   
2  telephone       226         1    999         0  NONEXISTENT           1.1   
3  telephone       151         1    999         0  NONEXISTENT           1.1   
4  telephone       307         1    999         0  NONEXISTENT           1.1   

  cons.price.idx cons.conf.idx euribor3m nr.employed   y                date  \
0         93,994         -36,4     4,857        5191  no       2-agosto-2019   
1         93,994         -36,4       NaN        5191  no  14-septiembre-2016   
2         93,994         -36,4     4,857        5191  no     15-febrero-2019   
3         93,994         -36,4       NaN        5191  no   29-noviembre-2015   
4         93,994         -36,4       NaN        5191  no       29-enero-2017   

   latitude  longitude                                   id_  
0    41.495    -71.233  089b39d8-e4d0-461b-87d4-814d71e0e079  
1    34.601    -83.923  e9d37224-cb6f-4942-98d7-46672963d097  
2    34.939    -94.847  3f9f49b5-e410-4948-bf6e-f9244f04918b  
3    49.041    -70.308  9991fafb-4447-451a-8be2-b0df6098d13e  
4    38.033   -104.463  eca60b76-70b6-4077-80ba-bc52e8ebb0eb

## Initial inspection

In [3]:
bank_df.shape

(43000, 24)

In [4]:
bank_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 43000 entries, 0 to 42999
Data columns (total 24 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Unnamed: 0      43000 non-null  int64  
 1   age             37880 non-null  float64
 2   job             42655 non-null  object 
 3   marital         42915 non-null  object 
 4   education       41193 non-null  object 
 5   default         34019 non-null  float64
 6   housing         41974 non-null  float64
 7   loan            41974 non-null  float64
 8   contact         43000 non-null  object 
 9   duration        43000 non-null  int64  
 10  campaign        43000 non-null  int64  
 11  pdays           43000 non-null  int64  
 12  previous        43000 non-null  int64  
 13  poutcome        43000 non-null  object 
 14  emp.var.rate    43000 non-null  float64
 15  cons.price.idx  42529 non-null  object 
 16  cons.conf.idx   43000 non-null  object 
 17  euribor3m       33744 non-null 

In [5]:
bank_df.describe(include="all").T

count unique                                   top   freq  \
Unnamed: 0      43000.0    NaN                                   NaN    NaN   
age             37880.0    NaN                                   NaN    NaN   
job               42655     11                                admin.  10873   
marital           42915      3                               MARRIED  25999   
education         41193      7                     university.degree  12722   
default         34019.0    NaN                                   NaN    NaN   
housing         41974.0    NaN                                   NaN    NaN   
loan            41974.0    NaN                                   NaN    NaN   
contact           43000      2                              cellular  27396   
duration        43000.0    NaN                                   NaN    NaN   
campaign        43000.0    NaN                                   NaN    NaN   
pdays           43000.0    NaN                                   NaN    NaN   
previous        43000.0    NaN                                   NaN    NaN   
poutcome          43000      3                           NONEXISTENT  37103   
emp.var.rate    43000.0    NaN                                   NaN    NaN   
cons.price.idx    42529     26                                93,994   7938   
cons.conf.idx     43000     26                                 -36,4   8020   
euribor3m         33744    309                                 4,857   2287   
nr.employed       43000     11                                5228,1  16980   
y                 43000      2                                    no  38156   
date              42752   1825                       28-febrero-2018     95   
latitude        43000.0    NaN                                   NaN    NaN   
longitude       43000.0    NaN                                   NaN    NaN   
id_               43000  43000  993bbbd6-4dbc-4a40-a408-f91f8462bee6      1   

                        mean           std      min        25%      50%  \
Unnamed: 0      20651.099047  11868.770913      0.0   10381.75  20642.5   
age                39.977112     10.437957     17.0       32.0     38.0   
job                      NaN           NaN      NaN        NaN      NaN   
marital                  NaN           NaN      NaN        NaN      NaN   
education                NaN           NaN      NaN        NaN      NaN   
default             0.000088       0.00939      0.0        0.0      0.0   
housing             0.535998      0.498708      0.0        0.0      1.0   
loan                 0.15562      0.362499      0.0        0.0      0.0   
contact                  NaN           NaN      NaN        NaN      NaN   
duration          257.739279    258.666033      0.0      102.0    179.0   
campaign            2.567233      2.772294      1.0        1.0      2.0   
pdays             962.330953    187.260394      0.0      999.0    999.0   
previous            0.174023      0.497366      0.0        0.0      0.0   
poutcome                 NaN           NaN      NaN        NaN      NaN   
emp.var.rate        0.077128      1.573898     -3.4       -1.8      1.1   
cons.price.idx           NaN           NaN      NaN        NaN      NaN   
cons.conf.idx            NaN           NaN      NaN        NaN      NaN   
euribor3m                NaN           NaN      NaN        NaN      NaN   
nr.employed              NaN           NaN      NaN        NaN      NaN   
y                        NaN           NaN      NaN        NaN      NaN   
date                     NaN           NaN      NaN        NaN      NaN   
latitude           36.856697      7.225948   24.396   30.61475   36.761   
longitude         -95.939067     16.752282 -124.997 -110.49425 -95.8995   
id_                      NaN           NaN      NaN        NaN      NaN   

                     75%      max  
Unnamed: 0      30930.25  41187.0  
age                 47.0     98.0  
job                  NaN      NaN  
marital              N

In [6]:
bank_df = bank_df.drop(columns="Unnamed: 0")

The column Unnamed: 0 was identified as redundant because it contains the same sequential values as the DataFrame’s default inde
Since this column does not provide additional information and duplicates the index, it was removed to clean the dataset and avoid unnecessary redundancy.

# Null values

In [7]:
bank_df.isnull().sum()

age               5120
job                345
marital             85
education         1807
default           8981
housing           1026
loan              1026
contact              0
duration             0
campaign             0
pdays                0
previous             0
poutcome             0
emp.var.rate         0
cons.price.idx     471
cons.conf.idx        0
euribor3m         9256
nr.employed          0
y                    0
date               248
latitude             0
longitude            0
id_                  0
dtype: int64

In [8]:
(bank_df.isna().sum())/len(bank_df)*100

age               11.906977
job                0.802326
marital            0.197674
education          4.202326
default           20.886047
housing            2.386047
loan               2.386047
contact            0.000000
duration           0.000000
campaign           0.000000
pdays              0.000000
previous           0.000000
poutcome           0.000000
emp.var.rate       0.000000
cons.price.idx     1.095349
cons.conf.idx      0.000000
euribor3m         21.525581
nr.employed        0.000000
y                  0.000000
date               0.576744
latitude           0.000000
longitude          0.000000
id_                0.000000
dtype: float64

For categorical columns with < 10% of null values we replace NaN with "unknown"

In [9]:
categorical_nan = ["job", "marital", "education", "default", "housing", "loan"]

for col in categorical_nan:
    bank_df[col] = bank_df[col].fillna("unknown")

In [10]:
bank_df.isna().sum()

age               5120
job                  0
marital              0
education            0
default              0
housing              0
loan                 0
contact              0
duration             0
campaign             0
pdays                0
previous             0
poutcome             0
emp.var.rate         0
cons.price.idx     471
cons.conf.idx        0
euribor3m         9256
nr.employed          0
y                    0
date               248
latitude             0
longitude            0
id_                  0
dtype: int64

Now we approach numerical columns with NaN
- age = 11.906977 % (~12%)
- cons.price.idx =  1.095349 % (~1%)
- euribor3m = 21.525581 % (~22%)
- date = 0.576744 % (~1%) 

### - age

In [11]:
bank_df["age"].describe()

count    37880.000000
mean        39.977112
std         10.437957
min         17.000000
25%         32.000000
50%         38.000000
75%         47.000000
max         98.000000
Name: age, dtype: float64

In [12]:
median_age = bank_df["age"].median()
bank_df["age"] = bank_df["age"].fillna(median_age)

In [13]:
bank_df["age"].isna().sum()

np.int64(0)

In [14]:
bank_df["age"].describe()

count    43000.000000
mean        39.741698
std          9.817735
min         17.000000
25%         33.000000
50%         38.000000
75%         46.000000
max         98.000000
Name: age, dtype: float64

The variable age had ~12% missing values.
Since age is a numerical variable and may contain outliers, the median was used for imputation in order to preserve the distribution and avoid distortion. After imputation, the central tendency remained stable.

### - cons.price.idx

In [15]:
bank_df["cons.price.idx"] = (bank_df["cons.price.idx"].str.replace(",", ".", regex=False)).astype("float64")
bank_df["cons.price.idx"].dtype

dtype('float64')

In [16]:
bank_df["cons.price.idx"].isnull().sum()

np.int64(471)

In [17]:
median_cpi = bank_df["cons.price.idx"].median()
bank_df["cons.price.idx"] = bank_df["cons.price.idx"].fillna(median_cpi)

In [18]:
bank_df["cons.price.idx"].isna().sum()

np.int64(0)

### - euribor3m

In [19]:
bank_df["euribor3m"].dtype

dtype('O')

In [20]:
bank_df["euribor3m"].describe()

count     33744
unique      309
top       4,857
freq       2287
Name: euribor3m, dtype: object

In [21]:
bank_df.columns

Index(['age', 'job', 'marital', 'education', 'default', 'housing', 'loan',
       'contact', 'duration', 'campaign', 'pdays', 'previous', 'poutcome',
       'emp.var.rate', 'cons.price.idx', 'cons.conf.idx', 'euribor3m',
       'nr.employed', 'y', 'date', 'latitude', 'longitude', 'id_'],
      dtype='object')

In [22]:
bank_df["date"].head()

0         2-agosto-2019
1    14-septiembre-2016
2       15-febrero-2019
3     29-noviembre-2015
4         29-enero-2017
Name: date, dtype: object

In [23]:
meses = {
    "enero": "January",
    "febrero": "February",
    "marzo": "March",
    "abril": "April",
    "mayo": "May",
    "junio": "June",
    "julio": "July",
    "agosto": "August",
    "septiembre": "September",
    "octubre": "October",
    "noviembre": "November",
    "diciembre": "December"
}
for esp, eng in meses.items():
    bank_df["date"] = bank_df["date"].str.replace(esp, eng, regex=False)

In [24]:
bank_df["date"] = pd.to_datetime(bank_df["date"], format="%d-%B-%Y", errors="coerce")

In [25]:
bank_df["date"].isna().sum()

np.int64(248)

In [26]:
bank_df["date"].dtype

dtype('<M8[ns]')

In [27]:
bank_df["contact_year"] = bank_df["date"].dt.year
bank_df["contact_year"].head()

0    2019.0
1    2016.0
2    2019.0
3    2015.0
4    2017.0
Name: contact_year, dtype: float64

In [28]:
bank_df["contact_year"] = bank_df["contact_year"].astype("Int64")
bank_df["contact_year"].dtype

Int64Dtype()

The date column contained Spanish month names, so months were translated to English and then parsed into datetime, enabling time-based analysis.

In [29]:
bank_df["euribor3m"] = (bank_df["euribor3m"].str.replace(",", ".", regex=False)).astype("Float64")
bank_df["euribor3m"].dtype

Float64Dtype()

In [30]:
bank_df["euribor3m"].isnull().sum()

np.int64(9256)

In [31]:
bank_df.groupby("contact_year")["euribor3m"].mean()

contact_year
2015    3.625715
2016    3.626521
2017    3.581983
2018    3.617693
2019    3.634085
Name: euribor3m, dtype: Float64

In [32]:
bank_df["euribor3m"] = bank_df["euribor3m"].fillna(bank_df.groupby("contact_year")["euribor3m"].transform("mean"))

In [33]:
bank_df["euribor3m"].isnull().sum()

np.int64(57)

Missing values in euribor3m (~22%) were imputed using the mean value within each contact year, preserving macroeconomic consistency across time.

In [34]:
bank_df.isna().sum()

age                 0
job                 0
marital             0
education           0
default             0
housing             0
loan                0
contact             0
duration            0
campaign            0
pdays               0
previous            0
poutcome            0
emp.var.rate        0
cons.price.idx      0
cons.conf.idx       0
euribor3m          57
nr.employed         0
y                   0
date              248
latitude            0
longitude           0
id_                 0
contact_year      248
dtype: int64

In [35]:
(bank_df["contact_year"].isna().sum())/len(bank_df)*100

np.float64(0.5767441860465117)

In [36]:
bank_df = bank_df.dropna(subset=["date"])

In [37]:
bank_df.isna().sum()

age               0
job               0
marital           0
education         0
default           0
housing           0
loan              0
contact           0
duration          0
campaign          0
pdays             0
previous          0
poutcome          0
emp.var.rate      0
cons.price.idx    0
cons.conf.idx     0
euribor3m         0
nr.employed       0
y                 0
date              0
latitude          0
longitude         0
id_               0
contact_year      0
dtype: int64

## Duplicate values

In [38]:
bank_df.duplicated().sum()

np.int64(0)

In [39]:
bank_df["id_"].duplicated().sum()

np.int64(0)

A check for fully duplicated rows was performed using duplicated().
No duplicate rows were found in the dataset.

This indicates that each row represents a unique observation.

Additionally, the id_ column — described as a unique identifier for each record — was examined for duplicate values. No duplicates were detected.

This confirms that each observation in the dataset represents a unique client interaction, and no further action regarding duplicates was required.

# Columns dtype

In [40]:
bank_df.head()

age        job  marital    education  default housing loan    contact  \
0  38.0  housemaid  MARRIED     basic.4y      0.0     0.0  0.0  telephone   
1  57.0   services  MARRIED  high.school  unknown     0.0  0.0  telephone   
2  37.0   services  MARRIED  high.school      0.0     1.0  0.0  telephone   
3  40.0     admin.  MARRIED     basic.6y      0.0     0.0  0.0  telephone   
4  56.0   services  MARRIED  high.school      0.0     0.0  1.0  telephone   

   duration  campaign  pdays  previous     poutcome  emp.var.rate  \
0       261         1    999         0  NONEXISTENT           1.1   
1       149         1    999         0  NONEXISTENT           1.1   
2       226         1    999         0  NONEXISTENT           1.1   
3       151         1    999         0  NONEXISTENT           1.1   
4       307         1    999         0  NONEXISTENT           1.1   

   cons.price.idx cons.conf.idx  euribor3m nr.employed   y       date  \
0          93.994         -36,4      4.857        5191  no 2019-08-02   
1          93.994         -36,4   3.626521        5191  no 2016-09-14   
2          93.994         -36,4      4.857        5191  no 2019-02-15   
3          93.994         -36,4   3.625715        5191  no 2015-11-29   
4          93.994         -36,4   3.581983        5191  no 2017-01-29   

   latitude  longitude                                   id_  contact_year  
0    41.495    -71.233  089b39d8-e4d0-461b-87d4-814d71e0e079          2019  
1    34.601    -83.923  e9d37224-cb6f-4942-98d7-46672963d097          2016  
2    34.939    -94.847  3f9f49b5-e410-4948-bf6e-f9244f04918b          2019  
3    49.041    -70.308  9991fafb-4447-451a-8be2-b0df6098d13e          2015  
4    38.033   -104.463  eca60b76-70b6-4077-80ba-bc52e8ebb0eb          2017

In [41]:
bank_df["id_"] = bank_df["id_"].astype("str")

In [42]:
bank_df["cons.conf.idx"] = (bank_df["cons.conf.idx"].str.replace(",", ".", regex=False)).astype("Float64")
bank_df["cons.conf.idx"].dtype

Float64Dtype()

In [43]:
bank_df["default"].value_counts()

default
0.0        33810
unknown     8939
1.0            3
Name: count, dtype: int64

In [44]:
bool_columns = ["housing", "loan"]
for col in bool_columns:
    bank_df[col] = bank_df[col].astype(bool)

In [45]:
bank_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 42752 entries, 0 to 42999
Data columns (total 24 columns):
 #   Column          Non-Null Count  Dtype         
---  ------          --------------  -----         
 0   age             42752 non-null  float64       
 1   job             42752 non-null  object        
 2   marital         42752 non-null  object        
 3   education       42752 non-null  object        
 4   default         42752 non-null  object        
 5   housing         42752 non-null  bool          
 6   loan            42752 non-null  bool          
 7   contact         42752 non-null  object        
 8   duration        42752 non-null  int64         
 9   campaign        42752 non-null  int64         
 10  pdays           42752 non-null  int64         
 11  previous        42752 non-null  int64         
 12  poutcome        42752 non-null  object        
 13  emp.var.rate    42752 non-null  float64       
 14  cons.price.idx  42752 non-null  float64       
 15  cons.co

In [46]:
bank_df["id_"].dtype

dtype('O')

In [56]:
bank_df.to_csv("../data/processed/bank_clean.csv", index=False)